ksp_plugin/part.hpp

```diff
@@ -16,49 +22,103 @@ namespace principia {
 namespace ksp_plugin {
 namespace internal_part {
 
+using base::IteratorOn;
 using base::not_null;
+using base::Subset;
 using geometry::Position;
 using geometry::Vector;
 using geometry::Velocity;
 using physics::DegreesOfFreedom;
-using quantities::Acceleration;
+using physics::DiscreteTrajectory;
+using quantities::Force;
 using quantities::Mass;
 
 // Corresponds to KSP's |Part.flightID|, *not* to |Part.uid|.  C#'s |uint|
 // corresponds to |uint32_t|.
 using PartId = std::uint32_t;
 
 // Represents a KSP part.
-template<typename Frame>
 class Part final {
  public:
-  Part(DegreesOfFreedom<Frame> const& degrees_of_freedom,
-       Mass const& mass,
-       Vector<Acceleration, Frame> const&
-           gravitational_acceleration_to_be_applied_by_ksp);
+  Part(PartId part_id, Mass const& mass);
 
-  DegreesOfFreedom<Frame> const& degrees_of_freedom() const;
-  Mass const& mass() const;
-  Vector<Acceleration, Frame> const&
-      gravitational_acceleration_to_be_applied_by_ksp() const;
+  virtual PartId part_id() const;
+
+  // Sets or returns the mass.  Event though a part is massless in the sense
+  // that it doesn't exert gravity, it has a mass used to determine its
+  // intrinsic acceleration.
+  virtual void set_mass(Mass const& mass);
+  virtual Mass const& mass() const;
+
+  // Clears, increments or returns the intrinsic force exerted on the part by
+  // its engines (or a tractor beam).
+  // TODO(phl): Keep track of the point where the force is applied.
+  virtual void clear_intrinsic_force();
+  virtual void increment_intrinsic_force(
+      Vector<Force, Barycentric> const& intrinsic_force);
+  virtual Vector<Force, Barycentric> const& intrinsic_force() const;
+
+  // Clears, sets or returns the degrees of freedom of the part.  A part that is
+  // not in the bubble (i.e., that belongs to a pile-up that is not in the
+  // bubble) doesn't have degrees of freedom.
+  virtual void clear_degrees_of_freedom();
+  virtual void set_degrees_of_freedom(
+      DegreesOfFreedom<Bubble> const& degrees_of_freedom);
+  virtual std::experimental::optional<DegreesOfFreedom<Bubble>> const&
+  degrees_of_freedom();
+
+  // This temporarily holds the trajectory followed by the part during the call
+  // to |PileUp::AdvanceTime| for the containing |PileUp|.  It read and cleared
+  // by |Vessel::AdvanceTime| for the containing |Vessel|.
+  virtual DiscreteTrajectory<Barycentric>& tail();
+  virtual DiscreteTrajectory<Barycentric> const& tail() const;
+
+  virtual bool tail_is_authoritative() const;
+  virtual void set_tail_is_authoritative(bool tail_is_authoritative);
+
+  // Requires |!is_piled_up()|.
+  virtual void set_containing_pile_up(IteratorOn<std::list<PileUp>> pile_up);
+
+  // An iterator to the |PileUp| containing |this|, if any.  Do not |Erase| this
```

Comment:
Let's not use `|this|` in comments, it's hard to read: "... to the containing |PileUp|, if any...".